### PR TITLE
Added Container and Routing debugging commands

### PR DESCRIPTION
**Pull request by mickaelandrieu**
This is sometimes very hard to debug the container or to be sure the routes are correctly mapped.

This is a first part of our new debugging toolkit.

## Routing, the right way
* Debug a route (and have complete informations)
dump-route
* Overview of all routes
dump-routes
* Overview of all routes with Controller (can break the view of some "low screen" resolutions)
dump-routes-controllers

## Container, the right way
* Search a service or a parameter
service-parameter-container
* List parameters
params
* List services
services

... and many more :)

### CODE REVIEW

**Comment by fkroockmann** [comment]
Nice job dude :)

**Comment by crouillon** [comment]
thanks @mickaelandrieu 
:+1:

**Comment by mickaelandrieu** [comment]
hmm scrutinizer complains, should I split my DebugContainerCommand into multiples commands ? 
/c  @eric-chau @crouillon

**Comment by eric-chau** [comment]
@mickaelandrieu what is your plan if you have to split it?

**Comment by eric-chau** [comment]
btw :+1:, really nice addition.

**Comment by mickaelandrieu** [comment]
@eric-chau ``DebugContainerServicesCommand`` and ``DebugContainerParametersCommand`` I guess ... I'm not very motivated to split theses commands are I work a lot to make them with the same behavior to the Symfony's one :/

File: Console/Command/DebugContainerCommand.php [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+        // sort so that it reads like an index of services
+        asort($serviceIds);
+
+        $name = $input->getArgument('name');
+        if ($name) {
+            $this->outputService($output, $name);
+        } else {
+            $this->outputServices($output, $serviceIds, $input->getOption('show-private'), $tag);
+        }
+    }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: new line is missing here =)

File: Console/Command/DebugContainerCommand.php [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        } elseif ($definition instanceof Alias) {
+            $alias = $definition;
+            $output->writeln(sprintf('This service is an alias for the service <info>%s</info>', (string) $alias));
+        } else {
+            // edge case (but true for "service_container", all we have is the service itself
+            $service = $definition;
+            $output->writeln(sprintf('<comment>Service Id</comment>   %s', $serviceId));
+            $output->writeln(sprintf('<comment>Class</comment>        %s', get_class($service)));
+        }
+    }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: New line's missing

File: Console/Command/DebugContainerCommand.php [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+                ++$optionsCount;
+            }
+        }
+        $name = $input->getArgument('name');
+        if ((null !== $name) && ($optionsCount > 0)) {
+            throw new \InvalidArgumentException('The options tags, tag, parameters & parameter can not be combined with the service name argument.');
+        } elseif ((null === $name) && $optionsCount > 1) {
+            throw new \InvalidArgumentException('The options tags, tag, parameters & parameter can not be combined together.');
+        }
+    }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: New line's missing

File: Console/Command/DebugContainerCommand.php [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            } elseif ($definition instanceof Alias) {
+                $alias = $definition;
+                $output->writeln(vsprintf($format, $this->buildArgumentsArray($serviceId, 'n/a', sprintf('<comment>alias for</comment> <info>%s</info>', (string) $alias), count($maxTags) ? array_fill(0, count($maxTags), '') : array())));
+            } else {
+                // we have no information (happens with "service_container")
+                $service = $definition;
+                $output->writeln(vsprintf($format, $this->buildArgumentsArray($serviceId, '', get_class($service), count($maxTags) ? array_fill(0, count($maxTags), '') : array())));
+            }
+        }
+    }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: New line's missing

File: Console/Command/DebugContainerCommand.php [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+    {
+        $arguments = array($serviceId);
+        foreach ($tagAttributes as $tagAttribute) {
+            $arguments[] = $tagAttribute;
+        }
+        $arguments[] = $scope;
+        $arguments[] = $className;
+
+        return $arguments;
+    }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: New line's missing

**Comment by mickaelandrieu** [comment]
Lines added. What do you think / spliting @eric-chau  ?

File: Console/Command/DebugContainerCommand.php [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        $this->containerBuilder = $this->getContainerBuilder();
+
+        if ($input->getOption('parameters')) {
+            $parameters = $this->getContainerBuilder()->getParameterBag()->all();
+            // Sort parameters alphabetically
+            ksort($parameters);
+            $this->outputParameters($output, $parameters);
+
+            return;
+        }
```
Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: Maybe you can do something like this to drop the complexity of your command class:

```php
protected function execute(InputInterface $input, OutputInterface $output)
{
    $this->validateInput($input);
    $this->containerBuilder = $this->getContainerBuilder();

if ($input->getOption('parameters')) {
    return $this->dumpContainerParameters();
}
```

And where ``::dumpContainerParameters`` is:

```php
protected function dumpContainerParameters()
{
    $parameters = $this->getContainerBuilder()->getParameterBag()->all();
    // Sort parameters alphabetically
    ksort($parameters);
    $this->outputParameters($output, $parameters);
}
```

What do you think about this? So your method ``::execute`` will be lighter.

Comment (eric-chau) [inline Console/Command/DebugContainerCommand.php]: Ofc you apply this logic to every options ^^

Comment (mickaelandrieu) [inline Console/Command/DebugContainerCommand.php]: yes, I will give a try this way.

According to Scrutinizer analysis, the worst methods are ``outputService`` and outputServices``. I don't think we can refactor a lot theses methods.

**Comment by mickaelandrieu** [comment]
closed in favor of a new ``DebugBundle`` => https://github.com/backbee/DebugBundle

**Pull request closed without merging**